### ISSUE TITLE
[v2] - Feature request - An ability to intercept k8s service

### ISSUE THREAD

**Issue by andreykrupskii** [telepresenceio/telepresence#1608]
First of all, I'm a big fan of telepresence as it's making the development process more convenient and saves my time. While I'm still not able to migrate to Telepresence 2 because of two problems. And this issue is about one of them. 

**Problem**: In our infrastructure micro-services besides the HTTP interface also communicate via message broker (rabbitmq). It was really easy to apply the previous version of Telepresence into our flow, because it just replaces an original container with the proxy, so the local instance is able to consume all messages from worker queues from the message broker. While Telepresence 2 injects sidecar container and the original instance stays alive, so messages from worker queues will be balanced between local and remote processes. 

**Feature request**: It would be a nice feature to be able to intercept traffic from a specific k8s service. So for our case, we would need just to scale the original deployment to zero and to start a local instance that will receive both HTTP traffic and message broker messages.

**Comment by AndreyKrupskii** [comment]
Another option is to extend `intercept` command with an option `--swap-original`, that will start the proxy container (as before) instead of injecting the sidecar.

**Comment by cujo** [comment]
This is must have feature for sure. My workaround for now is to replace image with stub while developing and then undo it:

set up
```
kubectl set image deployment/my-deployment my-container=ibmcom/pause:3.0
telepresence intercept my-service
```
tear down
```
telepresence quit
kubectl rollout undo deployment/my-deployment
```

**Comment by AndreyKrupskii** [comment]
@cujo, I've also tried to stub the original container while then I was faced with another issue: inability to intercept several container ports in parallel - https://github.com/telepresenceio/telepresence/issues/1607

**Comment by Splitside** [comment]
That would be a really useful feature!

**Comment by petergardfjall** [comment]
Wanted to pitch in that the current design of Telepresence v2 does not work well for the large class of applications that operate with work queue semantics.

For example where incoming requests get placed in a job queue later to be picked up by an available worker. This worked perfectly with Telepresence v1, which entirely swapped out the deployment for the local process. With v2, it's a fifty-fifty chance that a job gets handled by the (still-running) original pod inside the cluster, which makes the debugging value of telepresence very limited (in fact it almost adds confusion).

Having the option to either (1) _replace_ the deployment (as was done with `--swap-deployment` in v1) or (2) being able to intercept a _service_ (as suggested here), combined with the ability to scale down the backing workload to zero replicas, would also bring v2 to feature parity with v1 for this large class of applications.

**Comment by thallgren** [comment]
@petergardfjall there's a ticket requesting the ability to actually stop the running container #1646 and we're well aware that it would be helpful in a lot of use-cases where the  that level of intrusion doesn't matter (e.g. the whole cluster or namespace is used for testing only).

We've also done some work to allow less intrusive intercepts while still being able to handle message queues correctly. It requires header-propagation and filtering on the consumer side. Se [Telepresence RESTful API](https://www.getambassador.io/docs/telepresence/latest/reference/restapi/) for more info.

**Comment by petergardfjall** [comment]
@thallgren Good to hear that you are looking at this (from different angles). To me, I would be very happy if it would just be possible to replicate the behavior of `v1` where the in-cluster workload (deployment/replicaset containers) could be replaced. 

How difficult would it be to support the equivalent of a `--swap-deployment` flag to `telepresence intercept` (`telepresence intercept --replace` perhaps)? My (uneducated) guess is that it would just involve updating the workload to not only add the `traffic-agent`, but to _also remove the original container(s)_. Sounds like it could be simple (for someone familiar with the code base). Does that sound about right or are there additional complexities to it?

**Comment by thallgren** [comment]
@petergardfjall that sounds about right, but we must consider corner cases like: "what if someone wants to do a normal intercept after a `--replace`? Should that then re-inject the container into the pod, or is the `--replace` a permanent change? If it's the latter, isn't that something that should be configured on the workload (perhaps using an annotation)?

**Comment by petergardfjall** [comment]
> @petergardfjall that sounds about right, but we must consider corner cases like: "what if someone wants to do a normal intercept after a `--replace`? Should that then re-inject the container into the pod, or is the `--replace` a permanent change? If it's the latter, isn't that something that should be configured on the workload (perhaps using an annotation)?

I see. Could one make use of kubernetes rollbacks for that?

Edit: I mean to use rollbacks for `telepresence leave` on a (`--replace`)intercepted workload (and force a `leave` if a new `intercept` is done).

**Comment by 0BuRner** [comment]
Any news or workaround on this ?
Also face issue with celery workers

**Comment by thallgren** [comment]
@0BuRner no additions aside from the aforementioned Rest API at this time.

**Comment by jolo72** [comment]
I looked at the Rest API and it's not clear to me if it will help or not.  It looks interesting, but with Kafka, messages are assigned to partitions and partitions are assigned to specific clients, and only that client can read the message from Kafka.  If there's only 1 client (like in the old behavior with swap-deployment), then all partitions get assigned to the locally running instance.  Because of the partitioning, I don't think the Rest API could do anything with message headers to force a specific instance to process a particular message.

I am also voting for some way to re-introduce the v1 `swap-deployment` option.  If there's anything I can do to help with the understanding of the use case, let me know.

**Comment by petergardfjall** [comment]
The machinery for replacing the application container was implemented in https://github.com/telepresenceio/telepresence/pull/3330.

> This PR adds changes to the agent injector and traffic manager to enable the creation of intercepts that replace the application container.  The CLI flag is not implemented here.

If I understand things correctly the only thing missing to make this generally available is to add a `--replace` flag to the `telepresence intercept` CLI command.

I believe having `telepresence intercept --replace` solves the root problem of this issue ("replace rather than co-exist with existing application") and perhaps a few others like https://github.com/telepresenceio/telepresence/issues/1646.

This is the most highly anticipated feature judging by the number of :+1:s, so it would be interesting to know if there is an ETA for adding the  `--replace` flag to the cli.

Personally I'm still on `v1` and will probably stay on it until this gets added. But my impression is that there is very little work remaining. Would anyone care to comment?
